core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -75,197 +105,658 @@ object compress {
+                  _,
+                  _
+                )
+                ) =>
+              val secondsSince197001010000 =
+                unsignedToLong(header(4), header(5), header(6), header(7))
+              if (secondsSince197001010000 != 0)
+                modificationTimeCallback(Instant.ofEpochSecond(secondsSince197001010000))
+              val headerCrc32 = new CRC32()
+              val contentCrc32 = new CRC32()
+              headerCrc32.update(header)
+              val inflater = new Inflater(true)
+              Pull.output1(
+                streamAfterMandatoryHeader
+                  .through(_gunzip_skipOptionalExtraField(gzipFlag.fextra(flags), headerCrc32))
+                  .through(
+                    _gunzip_readOptionalStringField(
+                      gzipFlag.fname(flags),
+                      headerCrc32,
+                      fileNameCallback,
+                      "file name",
+                      fileNameBytesSoftLimit
+                    )
+                  )
+                  .through(
+                    _gunzip_readOptionalStringField(
+                      gzipFlag.fcomment(flags),
+                      headerCrc32,
+                      commentCallback,
+                      "file comment",
+                      fileCommentBytesSoftLimit
+                    )
+                  )
+                  .through(
+                    _gunzip_validateHeader((flags & gzipFlag.FHCRC) == gzipFlag.FHCRC, headerCrc32)
+                  )
+                  .through(
+                    _inflate(
+                      bufferSize = bufferSize,
+                      inflater = inflater,
+                      crc32 = Some(contentCrc32)
+                    )
+                  )
+                  .through(_gunzip_validateTrailer(contentCrc32, inflater)) ++
+                  Stream.suspend {
+                    inflater.end() // If F is narrowed to Applicative then inflater.end() can be bracketed rather than appended in an empty stream.
+                    Stream.empty
+                  }
+              )
+            case (
+                `gzipHeaderBytes`,
+                Array(
+                  `gzipMagicFirstByte`,
+                  `gzipMagicSecondByte`,
+                  compressionMethod,
+                  _,
+                  _,
+                  _,
+                  _,
+                  _,
+                  _,
+                  _
+                )
+                ) =>
+              Pull.raiseError(
+                new ZipException(s"Unsupported gzip compression method: $compressionMethod")
+              )
+            case _ =>
+              Pull.raiseError(new ZipException("Not in gzip format"))
           }
-          val pushed = abis.push(arr)
-          if (!pushed) throw NonProgressiveDecompressionException(bufferSize)
-        }
+        case None =>
+          Pull.raiseError(new EOFException())
+      }
+      .stream
+      .flatten
```

Comment:
Not to be That Guy™, but it's probably worth tossing together a quick benchmark to show that this implementation (which is inarguably nicer than what was there before!) doesn't cause any performance regressions. The previous implementation delegated all of the state management work to `GZIPOutputStream`, which has a reasonably well-optimized internal loop. This implementation is a lot nicer, but also does quite a bit more effect sequencing, which is usually very slow, and it's not entirely clear to me how much of it is in the hot path. Benchmarks would show for sure.

Reply:
I'll be happy to. Once the implementation review feedback has been completed I'll add benchmarks...